src/core/searcher.rs

```diff
@@ -89,6 +92,75 @@ impl Searcher {
         store_reader.get(doc_address.doc_id)
     }
 
+    /// Fetches documents from tantivy's store given a list of [`DocAddress`].
+    ///
+    /// This method is more efficient than calling [`doc`](Self::doc) multiple times, as it batches
+    /// overlapping requests to segments and blocks.
+    ///
+    /// The documents are returned in the same order as the `doc_addresses` slice.
+    pub fn docs<D: DocumentDeserialize + Clone>(
+        &self,
+        doc_addresses: &[DocAddress],
+    ) -> crate::Result<Vec<D>> {
+        // Ordering by segment ordinal allows reuse of a store reader for documents in the same
+        // segment.
+        let mut doc_addrs_by_segment_ord = HashMap::new();
+        for (idx, doc_address) in doc_addresses.iter().enumerate() {
+            doc_addrs_by_segment_ord
+                .entry(doc_address.segment_ord as usize)
+                .or_insert_with(HashMap::new)
+                .entry(doc_address)
+                .or_insert_with(Vec::new)
+                .push(idx);
+        }
+
+        let mut docs = Vec::with_capacity(doc_addresses.len());
+
+        for (segment_ord, doc_addr_to_idxs) in doc_addrs_by_segment_ord {
+            let store_reader = &self.inner.store_readers[segment_ord];
+
+            let doc_ids = doc_addr_to_idxs
+                .keys()
+                .map(|doc_address| doc_address.doc_id)
+                .collect();
+            let mut docs_in_segment: HashMap<u32, D> = store_reader.get_many(doc_ids)?;
+
+            for (doc_address, idxs) in doc_addr_to_idxs {
+                match idxs.as_slice() {
+                    [idx] => {
+                        debug_assert!(*idx < docs.capacity());
+
+                        // Safety: The index is valid because it was derived from `doc_addresses`.
+                        unsafe {
+                            *docs.get_unchecked_mut(*idx) =
```

Comment:
This produces a reference to uninitialized data which is instant UB, i.e. you need to either use `Vec<MaybeUninit<_>>` or initialize the elements using `docs.as_mut_ptr().add(*idx).write(..)`.